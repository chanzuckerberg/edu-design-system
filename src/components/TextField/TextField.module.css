@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #FIELD
\*------------------------------------*/

/**
 * 1) Consists of a label, form control, and an optional note about the field.
 */
.text-field {
  position: relative;
}

/**
 * Text field body
 */
.text-field__body {
  position: relative;
  margin-bottom: var(--eds-size-half);
}

/**
 * Textfield label 
 */
.text-field {
  .text-field--inverted & {
    color: var(--eds-theme-color-body-foreground-inverted);
  }
  .eds-is-disabled & {
    color: var(--eds-theme-color-disabled-foreground);
  }
}

/**
 * Text field label after
 * 1) Slot after the label that can be used to place components like a tooltip
 */
.text-field-after {
  display: inherit;
  margin-left: var(--eds-size-1);
}

/**
 * Text field input
 * 1) The actual form control
 */
.text-field__input {
  /**
     * Text field input within disabled field
     */
  .eds-is-disabled & {
    border-color: var(--eds-theme-color-disabled-border);
    background: var(--eds-theme-color-disabled-background);
    color: var(--eds-theme-color-disabled-foreground);
    cursor: not-allowed;
  }
}

/**
 * Text Field Within
 * 1) A slot to put arbitrary content that appears within
 *    the text field border to the right. 
 * 2) Typically used for buttons and icon buttons to enable
 *    things like show/hide password buttons 
 */
.text-field__input-within {
  position: absolute;
  right: var(--eds-size-2);
  top: 0;
  bottom: 0;
  display: grid;
  align-items: center;
  justify-content: center;
}
