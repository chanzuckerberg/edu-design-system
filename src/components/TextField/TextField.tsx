import clsx from 'clsx';
import React, { ChangeEventHandler, MouseEventHandler, ReactNode } from 'react';
import { useUID } from 'react-uid';
import styles from './TextField.module.css';
import FieldNote from '../../upcoming-components/FieldNote';
import Label from '../../upcoming-components/Label';
import TextInput from '../../upcoming-components/TextInput';
import Icon, { IconName } from '../Icon';

export interface Props {
  /**
   * Aria-describedby id string
   */
  ariaDescribedBy?: string;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Disables the field and prevents editing the contents
   */
  disabled?: boolean;
  /**
   * FieldNote
   * Used as helper text or error message
   */
  fieldNote?: string | ReactNode;
  /**
   * Toggles the visibility of the label. If hidden, the label text will still be accessible to assistive technologies
   */
  hideLabel?: boolean;
  /**
   * Screen reader text used for the button alongside the input field (i.e. show/hide password button)
   */
  fieldButtonScreenReaderText?: string;
  /**
   * Text used for the button alongside the input field (i.e. show/hide password button)
   */
  fieldButtonText?: string;
  /**
   * Function passed down from higher level component to trigger on click function of text field button
   */
  fieldButtonOnClick?: MouseEventHandler;
  /**
   * Name of SVG icon (i.e. chevron-down, minus, warning)
   */
  iconName?: IconName;
  /**
   * HTML id for the component
   */
  id?: string;
  /**
   * Gives a hint as to the type of data needed for text input
   */
  inputMode?:
    | 'text'
    | 'email'
    | 'url'
    | 'search'
    | 'tel'
    | 'none'
    | 'numeric'
    | 'decimal';
  /**
   * Node(s) that can be nested within the text field
   */
  inputWithin?: ReactNode;
  /**
   * Inverted variant for dark backgrounds
   */
  inverted?: boolean;
  /**
   * Error state of the form field
   */
  isError?: boolean;
  /**
   * HTML label text
   */
  label?: string;
  /**
   * Max number of characters for the text input
   */
  maxLength?: number;
  /**
   * Minimum value allowed for the input, if type is 'number'. When the input value matches this minimum, the minus button becomes disabled.
   */
  min?: number;
  /**
   * HTML name attribute for the input
   */
  name?: string;
  /**
   * Function that runs on change of the input
   */
  onChange?: ChangeEventHandler;
  /**
   * String for the optional label. By default it is '(optional)'
   */
  optionalLabel?: string;
  /**
   * Placeholder attribute for input. Note: placeholder should be used sparingly
   */
  placeholder?: string;
  /**
   * Toggles the form control's interactivity. When `readOnly` is set to `true`, the form control is not interactive
   */
  readOnly?: boolean;
  /**
   * Indicates that field is required for form to be successfully submitted
   */
  required?: boolean;
  /**
   * String for the required label to add additional information if needed.
   */
  requiredLabel?: string;
  /**
   * Title attribute on input
   */
  title?: string;
  /**
   * HTML type attribute, allowing switching between text, password, and other HTML5 input field types
   */
  type:
    | 'text'
    | 'password'
    | 'datetime'
    | 'datetime-local'
    | 'date'
    | 'month'
    | 'time'
    | 'week'
    | 'number'
    | 'email'
    | 'url'
    | 'search'
    | 'tel';
  /**
   * Value passed down from higher levels for initial state
   */
  value?: string | number;
  /**
   * Default value passed down from higher levels for initial state
   */
  defaultValue?: string | number;
}

/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * ```ts
 * import {TextField} from "@chanzuckerberg/eds";
 */
export const TextField = ({
  className,
  type = 'text',
  id,
  required,
  label,
  disabled,
  fieldNote,
  fieldButtonText,
  fieldButtonScreenReaderText,
  fieldButtonOnClick,
  inputWithin,
  name,
  value,
  defaultValue,
  onChange,
  placeholder,
  readOnly,
  iconName,
  inverted,
  hideLabel,
  maxLength,
  inputMode,
  ariaDescribedBy,
  isError,
  optionalLabel,
  requiredLabel,
  title,
  ...other
}: Props) => {
  const generatedId = useUID();
  const idVar = id || generatedId;

  const generatedAriaDescribedById = useUID();
  const ariaDescribedByVar = fieldNote
    ? ariaDescribedBy || generatedAriaDescribedById
    : undefined;

  const componentClassName = clsx(
    styles['text-field'],
    className,
    inverted && styles['text-field--inverted'],
    isError && styles['eds-is-error'],
    disabled && styles['eds-is-disabled'],
  );

  return (
    <div className={componentClassName}>
      <Label
        className={styles['text-field__label']}
        hideLabel={hideLabel}
        htmlFor={idVar}
        inverted={inverted}
        optionalLabel={optionalLabel}
        required={required}
        requiredLabel={requiredLabel}
        text={label}
      />

      <div className={styles['text-field__body']}>
        <TextInput
          aria-invalid={!!isError}
          ariaDescribedBy={ariaDescribedByVar}
          className={styles['text-field__input']}
          defaultValue={defaultValue}
          disabled={disabled}
          id={idVar}
          inputMode={inputMode}
          isError={isError}
          maxLength={maxLength}
          name={name}
          onChange={onChange}
          placeholder={placeholder}
          readOnly={readOnly}
          required={required}
          title={title}
          type={type}
          value={value}
          {...other}
        />
        {inputWithin && (
          <div className={styles['text-field__input-within']}>
            {inputWithin}
          </div>
        )}
        {iconName && (
          <Icon
            className={styles['text-field__icon']}
            name={iconName}
            purpose="decorative"
          />
        )}
      </div>
      {fieldNote && (
        <FieldNote
          className={styles['text-field__note']}
          id={ariaDescribedByVar}
          inverted={inverted}
          isError={isError}
        >
          {fieldNote}
        </FieldNote>
      )}
    </div>
  );
};
