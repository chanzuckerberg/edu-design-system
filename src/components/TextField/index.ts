export { TextField as default } from './TextField';
