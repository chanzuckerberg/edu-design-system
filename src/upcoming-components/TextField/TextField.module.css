@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #FIELD
\*------------------------------------*/

/**
 * Textfield label
 * 1) Consists of a label, form control, and an optional note about the field.
 */
.text-field {
  position: relative;

  .text-field--inverted & {
    color: var(--eds-theme-color-text-neutral-default-inverse);
  }
}

.text-field--inline {
  display: flex;
  align-items: center;
  gap: var(--eds-size-3);

  .text-field__note {
    margin-top: 0;
    margin-bottom: var(--eds-size-half);
  }
}

/**
 * Text field body
 */
.text-field__body {
  position: relative;
  margin-bottom: var(--eds-size-half);
}

/**
 * Text field label after
 * 1) Slot after the label that can be used to place components like a tooltip
 */
.text-field-after {
  display: inherit;
  margin-left: var(--eds-size-1);
}

/**
 * Text Field Within
 * 1) A slot to put arbitrary content that appears within
 *    the text field border to the right. 
 * 2) Typically used for buttons and icon buttons to enable
 *    things like show/hide password buttons 
 */
.text-field__input-within {
  position: absolute;
  right: var(--eds-size-2);
  top: 0;
  bottom: 0;
  display: grid;
  align-items: center;
  justify-content: center;
}
