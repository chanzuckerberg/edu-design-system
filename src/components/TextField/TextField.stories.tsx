import { Story, Meta } from '@storybook/react';
import React from 'react';

import { TextField, Props } from './TextField';
import Button from '../Button';

export default {
  title: 'Molecules/Forms/TextField',
  component: TextField,
} as Meta;

const Template: Story<Props> = (args) => <TextField {...args} />;

const InvertedTemplate: Story<Props> = (args) => (
  <div style={{ padding: '1rem', background: '#000' }}>
    <TextField {...args} />
  </div>
);

export const Default = Template.bind({});
Default.args = {
  fieldNote: 'This is a fieldnote.',
};

export const Inverted = InvertedTemplate.bind({});
Inverted.args = {
  inverted: true,
  fieldNote: 'This is a fieldnote.',
};

export const Error = Template.bind({});
Error.args = {
  isError: true,
  fieldNote: 'This is a fieldnote with an error.',
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Optional = Template.bind({});
Optional.args = {
  required: false,
  fieldNote: 'This is a fieldnote for an optional text field.',
};

export const HideLabel = Template.bind({});
HideLabel.args = {
  hideLabel: true,
};

export const InputWithin = () => {
  return (
    <div>
      <TextField
        type="text"
        inputWithin={
          <Button variant="bare" size="sm">
            Button
          </Button>
        }
      />
    </div>
  );
};
