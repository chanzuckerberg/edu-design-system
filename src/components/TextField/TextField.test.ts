import {generateSnapshots} from '@chanzuckerberg/story-utils';
import * as TextFieldStoryFile from './TextField.stories';

describe('<TextField />', () => {
  generateSnapshots(TextFieldStoryFile);
});
