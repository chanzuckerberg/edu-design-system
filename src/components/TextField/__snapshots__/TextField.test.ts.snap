// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Default story renders snapshot 1`] = `
<div
  class="text-field"
>
  <label
    class="label text-field__label"
    for="1"
  >
    Label
     
  </label>
  <div
    class="text-field__body"
  >
    <input
      aria-describedby="2"
      aria-invalid="false"
      class="text-input text-field__input"
      data-bootstrap-override="textinput"
      id="1"
      type="text"
      value=""
    />
  </div>
  <div
    aria-live="assertive"
    class="field-note text-field__note"
    id="2"
  >
    This is a fieldnote.
  </div>
</div>
`;

exports[`<TextField /> Disabled story renders snapshot 1`] = `
<div
  class="text-field eds-is-disabled"
>
  <label
    class="label text-field__label"
    for="8"
  >
    Label
     
  </label>
  <div
    class="text-field__body"
  >
    <input
      aria-invalid="false"
      class="text-input text-field__input eds-is-disabled"
      data-bootstrap-override="textinput"
      disabled=""
      id="8"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<TextField /> Error story renders snapshot 1`] = `
<div
  class="text-field eds-is-error"
>
  <label
    class="label text-field__label"
    for="5"
  >
    Label
     
  </label>
  <div
    class="text-field__body"
  >
    <input
      aria-describedby="6"
      aria-invalid="true"
      class="text-input text-field__input eds-is-error"
      data-bootstrap-override="textinput"
      id="5"
      type="text"
      value=""
    />
  </div>
  <div
    aria-live="assertive"
    class="field-note text-field__note eds-is-error"
    id="6"
  >
    <svg
      class="icon field-note__icon"
      fill="currentColor"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="7"
      >
        error
      </title>
      <use
        xlink:href="test-file-stub#cancel"
      />
    </svg>
    This is a fieldnote with an error.
  </div>
</div>
`;

exports[`<TextField /> HideLabel story renders snapshot 1`] = `
<div
  class="text-field"
>
  <label
    class="label text-field__label u-is-vishidden"
    for="12"
  >
    Label
     
  </label>
  <div
    class="text-field__body"
  >
    <input
      aria-invalid="false"
      class="text-input text-field__input"
      data-bootstrap-override="textinput"
      id="12"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<TextField /> InputWithin story renders snapshot 1`] = `
<div>
  <div
    class="text-field"
  >
    <label
      class="label text-field__label"
      for="14"
    >
      Label
       
    </label>
    <div
      class="text-field__body"
    >
      <input
        aria-invalid="false"
        class="text-input text-field__input"
        data-bootstrap-override="textinput"
        id="14"
        type="text"
        value=""
      />
      <div
        class="text-field__input-within"
      >
        <button
          class="clickable-style clickable-style--sm clickable-style--icon clickable-style--brand button button--icon"
          data-bootstrap-override="clickable-style-icon"
          type="button"
        >
          Button
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TextField /> Inverted story renders snapshot 1`] = `
<div
  style="padding: 1rem; background: rgb(0, 0, 0);"
>
  <div
    class="text-field text-field--inverted"
  >
    <label
      class="label text-field__label label--inverted"
      for="3"
    >
      Label
       
    </label>
    <div
      class="text-field__body"
    >
      <input
        aria-describedby="4"
        aria-invalid="false"
        class="text-input text-field__input"
        data-bootstrap-override="textinput"
        id="3"
        type="text"
        value=""
      />
    </div>
    <div
      aria-live="assertive"
      class="field-note text-field__note field-note--inverted"
      id="4"
    >
      This is a fieldnote.
    </div>
  </div>
</div>
`;

exports[`<TextField /> Optional story renders snapshot 1`] = `
<div
  class="text-field"
>
  <label
    class="label text-field__label"
    for="10"
  >
    Label
     
    <span
      class="label__flag"
    >
      (optional)
    </span>
  </label>
  <div
    class="text-field__body"
  >
    <input
      aria-describedby="11"
      aria-invalid="false"
      class="text-input text-field__input"
      data-bootstrap-override="textinput"
      id="10"
      type="text"
      value=""
    />
  </div>
  <div
    aria-live="assertive"
    class="field-note text-field__note"
    id="11"
  >
    This is a fieldnote for an optional text field.
  </div>
</div>
`;
